### PR TITLE
feature: Delay recompilation until no file changes in last 1s

### LINKED ISSUE #191: Introduce a delay (e.g. 1s) for recompilation upon file change
Right now it is being triggered a lot, especially by tmp editor files, so this would probably save a lot of unnecessary recompilations.

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

Previously, after any file change we would recompile the web app. Now, it will defer recompilation until no change was detected in the last second.

Closes #191

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by shayneczyzewski** [comment]
Open question: Is my use of Left and Right canonical here? I feel like Either being Right-based may mean I should switch the two perhaps? Or is that only useful in Error/monadic settings and not really in this context?

**Comment by shayneczyzewski** [comment]
Side note: I believe the only possible race condition that could exist is as follows:
- Timer expires and filesystem event happen at the exact same time
- Timer wins, but we consume the `readChan` before `race` can cancel the thread (so net result is timer wins but we "lost" an event)
In my mind, the recompile should still pick it up, but is that 100% guaranteed?

File: waspc/cli/Wasp/Cli/Command/Watch.hs [hunk 1]
```diff
@@ -29,25 +31,37 @@ import qualified Wasp.Lib
...
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
-      event <- readChan chan
-      let eventTime = FSN.eventTime event
-      if eventTime < lastCompileTime
-        then -- If event happened before last compilation started, skip it.
-          listenForEvents chan lastCompileTime
-        else do
+    listenForEvents :: Chan FSN.Event -> UTCTime -> Bool -> IO ()
```
Comment (Martinsos) [inline waspc/cli/Wasp/Cli/Command/Watch.hs:43]: Looking good! Look at the algorithm, it seems to me like it should work ok.
I did find it a bit hard to follow though, in the sense that control flow is a bit complex -> depending on value of `recompileAfterPreemptibleDelay` and on the value returned by the `race`, there are different things that can happen. Due to `recompileAfterPreemptibleDelay` the function effectively operates in two different modes. I wonder if we could somehow split the code into multiple functions to make this a bit more explicit / simpler?

I will give it a quick try here, in pseudocode:
```
- listenForEvents (fn)
  - wait for event
  - if an old event
    - continue from start (listenForEvents)
    else
    - race (wait for another event) (1s delay)
    - if another event && is new enough
      - go back to `race` above
      else
      - recompile
      - go to start (listenForEvents)
```   
     
It is the same thing really, but I think it is a bit clearer regarding the control flow?
As for repeating the race again and again, I would just create another function called `waitUntilNoEvents` or smth like that (there is probably a better name) that takes duration and basically blocks until no events happened for that duration -> it would be this part of logic that calls `race` again and again.
And then that would be called from the `listenForEvents`.

So that would then be

```
- listenForEvents (fn)
  - wait for event
  - if an old event
    - continue from start (listenForEvents)
    else
    - waitUntilNoEvents 1s
    - recompile
    - go to start (listenForEvents)

- waitUntilNoEvents (fn)
  - race (wait for another event) (1s delay)
    - if another event && is new enough
      - go to start (waitUntilNoEvents)
      else
      - done
```   

Also I see that I used 2 times the concept of "event is old/young" in pseudocode -> maybe that could also be made into a separate function? Not sure though.

File: waspc/cli/Wasp/Cli/Command/Watch.hs [hunk 2]
```diff
@@ -29,25 +31,37 @@ import qualified Wasp.Lib
...
+    listenForEvents chan lastCompileTime recompileAfterPreemptibleDelay = do
+      let microsecondsPerSecond = 1000000
+      res <-
+        if recompileAfterPreemptibleDelay
+          then -- Recompile after 1s, unless we get another file change before then.
+            race (readChan chan) (threadDelay $ 1 * microsecondsPerSecond)
+          else -- Just wait for next change.
+            Left <$> readChan chan
+      case res of
+        Left event -> do
```
Comment (Martinsos) [inline waspc/cli/Wasp/Cli/Command/Watch.hs]: Usage of Left and Right -> sure, it looks ok! This is how `race` uses them anyway, so I think you did all good here. `Either` is often used for storing error in `Left` and value in `Right`, main reason being that `Either` is functor in the `Right` value, but here we are using it for different purpose and that is ok.

Comment (Martinsos) [reply waspc/cli/Wasp/Cli/Command/Watch.hs]: Aha, got what you mean by swapping Left and Right -> again, I think it doesn't matter, it is up to you completely. Only thing you might want to keep in mind is this Functor thing, but in this case it really doesn't matter. It makes a bit of sense to me to have Right as wanted behaviour, while Left is kind of unwanted -> and since we want delay to succeed, it makes just a bit more sense to me to keep it in Right, as you did initially. But again, not really important here.

Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs]: Nice, thanks for confirming that! I did swap in a cleanup refactor out of "just in case" :D , but good to know it was not strictly necessary!

**Comment by Martinsos** [comment]
> Side note: I believe the only possible race condition that could exist is as follows:
> 
> * Timer expires and filesystem event happen at the exact same time
> * Timer wins, but we consume the `readChan` before `race` can cancel the thread (so net result is timer wins but we "lost" an event)
>   In my mind, the recompile should in practice still pick it up, but is that 100% guaranteed? Is it 100% avoidable with APIs that do not require an ack when getting `readChan` events?
> 
> Update: It appears the semantics of `race` + `cancel` should handle this case fine and we do not have a race there: https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html
> 
> > Note that cancel will not terminate until the thread the Async refers to has terminated. This means that cancel will block for as long said thread blocks when receiving an asynchronous exception.
> 
> The attempt to `cancel` the `readChan chan` that successfully read from the channel but lost the race will be a no-op, as that thread will have logically completed already too by the time it was attempted to be cancelled.

Hm very good point! So what happens then if "both" win at the same time? Which one is the winner then, and what happens to that event that was read -> maybe you explained this but I didn't quite get it.

**Comment by Martinsos** [comment]
One thing that would be interesting here is adding some tests! Would you be up for that? I think it is a good opportunity to play with that a bit -> with testing IO. I wouldn't want us to loose a lot of time here though, but if we can put a bit of time into it and write some successful IO tests, that could be a cool learning opportunity for all of us + we will have this piece of code tested.

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:43]: Thanks for the thoughts on trying to simplify the logic. I will review this sketch and test it out in code and see how it looks/feels! Appreciate it 👍🏻

**Comment by shayneczyzewski** [comment]
> One thing that would be interesting here is adding some tests! Would you be up for that? I think it is a good opportunity to play with that a bit -> with testing IO. I wouldn't want us to loose a lot of time here though, but if we can put a bit of time into it and write some successful IO tests, that could be a cool learning opportunity for all of us + we will have this piece of code tested.

Sounds like fun, I can give it a few hours and see how much progress I can make on it!

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 1 above)
Comment (Martinsos) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:43]: Np!
Btw while reviewing the code I remembered one thing from the Clean Code book -> it says that functions taking Bool as an argument are always suspects for doing too much :D. And it refers to the case when Bool is used to control the flow of the function. Recommended "fix" is to try and split function into multiple functions. So that immediately caught my eye, and I think is a useful rule (if not followed blindly of course).

Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:43]: Yeah, it is a good call to not allow the function to get too split brained for sure!

**Comment by Martinsos** [comment]
> > Hm very good point! So what happens then if "both" win at the same time? Which one is the winner then, and what happens to that event that was read -> maybe you explained this but I didn't quite get it.
> 
> I think if they both win, but we declare the timer the winner, the net result of reading the channel but losing the event should be the same -> recompile. Worst case it was an "old event" but we were already waiting to recompile anyways. So I think as long as the tie result is to do the recompile we were wanting/waiting to do, we will still get that file and the desired result. Lmk though if that does not make sense or sounds off.

Yeah that makes sense!! I am not sure why I thought it would be a problem hah :D.
Btw I believe there is also something like `peekEvent` instead of `readEvent` in case that we wanted to react on new event but not consume it -> however I don't think that is needed, what you said here makes perfect sense.

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 1 above)
Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs]: Refactor here: https://github.com/wasp-lang/wasp/pull/403/commits/ccca4fe67717314a03e83407f8004a9344f57d4f Agree much easier to follow separated out, thnx for the suggestion, will keep my eyes peeled for those opportunities more moving forward :D

**Comment by shayneczyzewski** [comment]
I have opened a new issue for adding test support to cli @Martinsos https://github.com/wasp-lang/wasp/issues/405

I think this squashed PR is in pretty good shape, but please let me know if anything jumps out at you upon final review. Thanks again for the good suggestion to split into two functions! 👍🏻  I think this was a fun issue to work on and think about.

File: waspc/cli/Wasp/Cli/Command/Watch.hs [hunk 3]
```diff
@@ -40,15 +43,41 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
...
+        then -- Ignore delayed/stale events older than our last compile time.
           listenForEvents chan lastCompileTime
         else do
+          -- Recompile, but only after a 1s period of no new events.
+          waitUntilNoNewEvents chan lastCompileTime 1
           currentTime <- getCurrentTime
           recompile
           listenForEvents chan currentTime
 
+    -- Blocks until no new events are recieved for a duration of `secondsToDelay`.
```
Comment (Martinsos) [inline waspc/cli/Wasp/Cli/Command/Watch.hs:56]: Nice comment!
In Haskell, there is a tool called Haddock that parses comments and creates documentation from them. It introduces some special syntax to comments, you can find it in Haddock docs.
This is usually important for libraries and similar, it is not that important for our codebase, but still one useful thing is to use 
```
-- | This refers to the code below.
```

and

```
-- ^ This refers to code that is to the left (above?)
```

So it would be:
```hs
-- | Blocks until no new events are recieved for a duration of `secondsToDelay`.
-- Consumes any new events during an active timer window and then restarts wait.
-- If a stale event comes in during an active timer window, we immediately
-- return control to the caller.
```

File: waspc/cli/Wasp/Cli/Command/Watch.hs [hunk 4]
```diff
@@ -40,15 +43,41 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
...
+    -- return control to the caller.
+    waitUntilNoNewEvents :: Chan FSN.Event -> UTCTime -> Int -> IO ()
+    waitUntilNoNewEvents chan lastCompileTime secondsToDelay = do
+      eventOrDelay <- race (readChan chan) (threadDelaySeconds secondsToDelay)
+      case eventOrDelay of
+        Left event -> do
+          if isStaleEvent event lastCompileTime
+            then return ()
+            else -- We have a new event, restart waiting process.
+              waitUntilNoNewEvents chan lastCompileTime secondsToDelay
```
Comment (Martinsos) [inline waspc/cli/Wasp/Cli/Command/Watch.hs:68]: You can do
```suggestion
          unless (isStaleEvent event lastCompileTime) $
            -- We have a new event, restart waiting process.
            waitUntilNoNewEvents chan lastCompileTime secondsToDelay
```

There is also an opposite function, `when`.

File: waspc/cli/Wasp/Cli/Command/Watch.hs [hunk 5]
```diff
@@ -40,15 +43,41 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
...
+        Right () -> return ()
+
+    isStaleEvent :: FSN.Event -> UTCTime -> Bool
+    isStaleEvent event lastCompileTime =
+      let eventTime = FSN.eventTime event
+       in eventTime < lastCompileTime
```
Comment (Martinsos) [inline waspc/cli/Wasp/Cli/Command/Watch.hs:74]:
```suggestion
    isStaleEvent event lastCompileTime = FSN.eventTime event < lastCompileTime
```

**Review by Martinsos — APPROVED** [review]
Great :)! I left couple of comments but all very minor, therefore I already approved -> resolve those however you feel fit and you can merge! Nice job, and it does look like a fun piece of code :).

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 3 above)
Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:56]: Cool, thanks for the tip on Haddock! I will update that 👍🏻

Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:56]: Hmm, I tried updating it, but when I save the formatter removes the comment all-together when I use the Haddock style (both `-- |` and `-- ^`). I wonder if that is because it is a `where` function?

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 4 above)
Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:68]: Ah cool, thanks for bringing this up! I was actually originally considering using `unless` but I admittedly wasn't exactly sure how it handled the alternative case from the docs. So when the `unless` fails in an `IO ()` case it defaults to `return ()`? That makes sense. Pretty handy! This does look cleaner! Thnx

```
      case eventOrDelay of
        Left event -> do
          unless (isStaleEvent event lastCompileTime) $
            -- We have a new event, restart waiting process.
            waitUntilNoNewEvents chan lastCompileTime secondsToDelay
        Right () -> return ()
```

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 5 above)
Comment (shayneczyzewski) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:74]: Ooff good call 🤦🏻 👍🏻

**Comment by shayneczyzewski** [comment]
> Great :)! I left couple of comments but all very minor, therefore I already approved -> resolve those however you feel fit and you can merge! Nice job, and it does look like a fun piece of code :).

Thanks @Martinsos! I was able to update 2 of the 3 suggestions, for some reason the Haddock doc one didn't seem to stick in the `where` function context. I appreciate the tip on `unless`! 👍🏻 I will do a final pass, squash, and then merge tomorrow. I try not to do that in the evening when the brain is a bit foggy 😄  Thanks again for the review and great comments

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 3 above)
Comment (Martinsos) [reply waspc/cli/Wasp/Cli/Command/Watch.hs:57]: Hmmmm possible, although I would expect it to work as a function header. Ok nm then!

File: waspc/cli/Wasp/Cli/Command/Watch.hs (hunk 4 above)
Comment (Martinsos) [reply waspc/cli/Wasp/Cli/Command/Watch.hs]: Yes, basically it does exactly what your previous code was doing, calls return ()! And same goes for `when`.
